### DISCUSSION TITLE
Feature Request: Add an easy way to supply two different file paths as parameters

### DISCUSSION THREAD

**Discussion by LennartCode** [Taitava/obsidian-shellcommands#378] [Ideas]
## Background

I want do diff two files from my vault using VSCode and therefore use the plugin to run `code --diff <file 1 absolute path> <file 1 absolute path>`.
This works excellently but with the small downside that I can only read one file path automatically, currently using `{{file_path:absolute}}` - as far as I know, please tell me if there is a more convenient way. The other parameter is supplied using a custom variable for which I have to paste the second file path by hand, which in turn requires leaving obsidian and thus defeats the whole purpose of automating the diff in the first place.

## Request

I would love for the plugin to be able to automatically read two paths from within Obsidian to spare me the hustle of changing applications manually.

## Suggestions

The greatest but probably most specific addition would be the ability to right click a file and store the absolute path to a variable. I don't know how many use cases there are for others though.

Another great alternative would be, if the plugin was able to read the path other files that are opened 'to the right', maybe using an index. In this scenario, I would open the two files to diff side by side, run the shell command, which takes the first two panes opened next to each other and VSCode would open the diff of those files.

## Thanks

This plugin is amazing and saves me tons of time. Thank you very much for developing such a powerful and easy to use tool!

**Comment by Taitava** [comment]
Hi! Thank you for your suggestion and kind feedback! 🙂

I do agree that it's unfortunately limiting that the plugin does not currently provide a way to retrieve properties from all the open files.

> Another great alternative would be, if the plugin was able to read the path other files that are opened 'to the right', maybe using an index.

Hmm, yeah, it could make sense to be able to access files to the left or right of the currently focused file. I'd need to figure out the syntax for it. It can get quite complex, as Obsidian is very flexible by its user interface:
- opened files reside in tabs
- one pane can contain multiple tabs
- one window can contain multiple panes
- there can be multiple windows open

So, how to come up with a syntax that could allow different scenarios? For example:
- Get file paths of all open files from all panes and all windows.
- Get file paths from the currently focused pane only.
- Get a file path from the next tab to the right (like in your case), or the previous tab to the left.
- Get file paths from a pane in different window/pane that is not focused.
- And of course the property could be something else than a path, too, e.g. [`{{tags}}`](https://publish.obsidian.md/shellcommands/Variables/%7B%7Btags%7D%7D).

## Variable scoping

There should be some kind of scoping mechanism: The normal variable `{{file_path:absolute}}` could be preceded like this: `{{tab(+1).file_path:absolute}}` to get the path of the file to the right of the currently focused tab. Use `+2` to pick the second tab to the right. Use `-1` to get a tab to the left. Use `0` without a preceding sign to get the absolute leftmost tab. This is just a very quick idea, so the syntax might change.

Also, to support a dynamic amount of tabs, there could be a scope that could provide access to multiple files at once: `{{tabs.file_path:absolute}}->join((, ))` could give paths from all open files. As it's providing a list of values, there needs to be a stringifier after the variable: `->join((, ))`. The result of joining could be e.g. `/home/me/Obsidian/Note.md, /home/me/Obsidian/AnotherNote.md, /home/me/Obsidian/ThirdNote.md`. Or, instead of `->join(())`, you could use e.g. `->count(())` to just get a number of opened files if it's useful somewhere.

I've been thinking about the ability to access multiple files every now and then, but haven't yet come up with a clear plan for it. I'll keep thinking about it, perhaps it will clear up to me some day.

## One workaround for now

What you can do currently, is that you can use the [File menu event](https://publish.obsidian.md/shellcommands/Events/File+menu) to have your shell command appear in the menu if you right click on a file. Then, in your shell command you can use [{{event_file_path}}](https://publish.obsidian.md/shellcommands/Variables/%7B%7Bevent_file_path%7D%7D) to get the right clicked file's path, and [{{file_path}}](https://publish.obsidian.md/shellcommands/Variables/%7B%7Bfile_path%7D%7D) to get the currently focused file's path (= the one which has a caret). These can be two different files.

I hope this provided some help to the situation. 🙂

**Reply by LennartCode** [reply]
Thank you very much for your thoughts!

All of your suggestions for how to approach this feature request seem awesome and would not only solve my problem but also open great possibilities for others to do more productive stuff, I assume :)

For me though, the workaround you suggested is more than sufficient though. This is exactly what I would have wished for and definitely solves my problem. 🎉

So thank you very much! :D

**Reply by Taitava** [reply]
You are welcome! 😎 And thank you for your kind words! 🌞
